### DISCUSSION TITLE
disable the weekly dev build

### DISCUSSION THREAD

**Discussion by FeynmanZhou** [notaryproject/notation#372] [General]
Hi team,

After analyzing the Notation GitHub releases download data, I find the following problems:

- The downloads decreased obviously starting from the first weekly build release
- Each weekly dev build is less than 5 downloads, no users are really using it
- There are too many weekly build releases on the GitHub release page, which confused the new users to choose
- All releases are `pre-release`, users are hard to find the recommended release from the [homepage](https://github.com/notaryproject/notation), people may concern about the `pre-release` status

![image](https://user-images.githubusercontent.com/40452856/194968885-867dbbef-39b9-4018-a1f6-72c9027bf373.png)
![image](https://user-images.githubusercontent.com/40452856/194968939-ef1725d3-853f-4883-82e6-5e9ca8808fdf.png)

According to [SemVer](https://semver.org/), we can convert a `pre-release` to an official release even if it is still in `alpha` or `v0.x.0`. I propose enabling the official release start from Alpha.4 and disabling the weekly dev build now.

**Comment by dtzar** [comment]
I'm good to move from pre-release to release.  However, I'd still like to keep 1 or 2 weekly dev builds and [auto-remove the dev builds](https://github.com/notaryproject/notation/issues/369) when a new one comes out.

**Reply by FeynmanZhou** [reply]
Follow up on this proposal:
1. Notation v0.11.0-Alpha.4 was set to an standard release and can be easily found from the [homepage](https://github.com/notaryproject/notation).

2. @dtzar  Keeping only 1 latest dev build release sounds good to me. @JeyJeyGao  Is this request doable in GH actions?
But from the latest two releases' download stats (5 and 0 downloads), a less of people were really using the dev build. The release rhythm is also fast due to the beta.1 and rc.1 release will come within around 3 weeks. So I still prefer to disable it temporally. 
![image](https://user-images.githubusercontent.com/40452856/197676567-4b42a5bf-6ef2-475f-b997-b56e0741b9a4.png)

cc @yizha1 @iamsamirzon @vaninrao10 @notaryproject/notation-release-managers

**Reply by junczhu** [reply]
Hi @FeynmanZhou, you can set a GitHub Action to manage latest dev build release.

[GitHub Action Event Trigger including *Release*](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#release:~:text=%5Bpublished%5D-,release,-Webhook%20event%20payload)

For example, you can run a workflow when a new dev release has been `published`.

```yaml
on:
  release:
    types: [published]
```
According to the GitHub Docs, we can [using the GitHub CLI on a runner](https://docs.github.com/en/actions/examples/using-the-github-cli-on-a-runner#understanding-the-example:~:text=Search%20GitHub%20Docs-,Using%20the%20GitHub%20CLI%20on%20a%20runner,-In%20this%20article).

And [using GitHub CLI to list releases](https://docs.github.com/en/rest/releases/releases#list-releases:~:text=and%20release%20assets.-,List%20releases,-Works%20with%20GitHub) / [delete a release](https://docs.github.com/en/rest/releases/releases#delete-a-release:~:text=%7D%0A%20%20%5D%0A%7D-,Delete%20a%20release,-Works%20with%20GitHub)

More Info:

Here is a [sample GitHub Action](https://github.com/marketplace/actions/delete-older-releases) from Marketplace. This action deletes older releases of given repo.
And here is the [source code](https://github.com/dev-drprasad/delete-older-releases) of the GitHub Action above, which is under MIT LICENSE

**Comment by JeyJeyGao** [comment]
Do we really need weekly dev release now? Can we stop it if no one uses it?

**Reply by dtzar** [reply]
I believe it would be a good idea to leave only the latest weekly dev build in order to help with automated AND manual testing to find problems.  IF notation does eventually consistently move to a release every 2 weeks, then at that point yes I think it's a good consideration to remove entirely.

**Reply by dtzar** [reply]
It doesn't matter how many people download the dev release as long as people can see the major release easily.  For instance - it only takes one person downloading and trying out the dev release to find a problem (or potentially realize we fixed something that week they needed).

Side note: @FeynmanZhou - I recommend removing any numbers of how many people download our releases in general.  @toddysm

**Reply by SteveLasker** [reply]
Regular releases forces a discipline of always being ready to ship, having stability in main, and having the opportunity to validate a released set of bits. 
Is there a downside to releasing weekly?

**Reply by vaninrao10** [reply]
> Regular releases forces a discipline of always being ready to ship, having stability in main, and having the opportunity to validate a released set of bits. Is there a downside to releasing weekly?

I dont see any downside. Not sure if others see anything here. We can definitely continue to do this weekly dev build.

**Reply by toddysm** [reply]
While I see where @dtzar, @SteveLasker, and @vaninrao10 are coming from I have a few questions:
- What is the value to keep older weekly "releases"? Do we expect somebody to go back and use those? If not, they are just polluting the releases and may add to confusion amongst customers (see below) or make it hard to find the right release.
- Who are the target users of the weekly "releases"? If those are developers then they can build from source, if there are other users we expect then we should clarify.
- With so many "releases", how will we make it for users to just find the one they need (supposedly the latest stable release)?

**Reply by SteveLasker** [reply]
great questions 
I don’t think we need lots of weekly releases maintained. We should be having frequent builds, so these are more interim

Building from source sounds far easier than it is if you don’t have the correct environment setup. curl is far easier than make, debug golang versions

Latest 3 is probably fine. Once we have good test coverage, fewer is probably fine.

**Reply by dtzar** [reply]
- As mentioned: I recommend only keeping the latest weekly dev build, so at any point there is only one dev build available from the last week.  This will make it easy to find the actual release.
- The target users are any users or developers who don't want to have to build the binary on their own to test out the latest improvements.  The target is also the development team to be able to run that released binary with e2e test suites and also be on the same page with what specific bits are being manually tested.
- Answered above.

**Reply by FeynmanZhou** [reply]
+1 to keep the latest weekly dev build. The elder dev build does not make sense for real testing cases as it doesn't include the latest PRs in it. 

@dtzar Everyone can easily get the download stats from an open-source project, it's transparent on the Internet. Do we need to remove the screenshot above?

**Reply by FeynmanZhou** [reply]
@JeyJeyGao I will turn this discussion into a GitHub issue. Can you update the release workflow accordingly?

**Reply by JeyJeyGao** [reply]
The old release deleting logic added in this PR: https://github.com/notaryproject/notation/pull/449
